go/signedexchange/signedexchange_test.go

```diff
@@ -0,0 +1,169 @@
+)
+
+type zeroReader struct{}
+
+func (zeroReader) Read(b []byte) (int, error) {
+	for i := range b {
+		b[i] = 0
+	}
+	return len(b), nil
+}
+
+func TestSignedExchange(t *testing.T) {
+	u, _ := url.Parse("https://example.com/")
+	header := http.Header{}
+	header.Add("Content-Type", "text/html; charset=utf-8")
+	i, err := signedexchange.NewInput(u, 200, header, []byte(payload), 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	i.AddSignedHeadersHeader("Content-Type", "Content-Encoding", "MI")
+
+	now := time.Date(2018, 1, 31, 17, 13, 20, 0, time.UTC)
+	certs, err := signedexchange.ParseCertificates([]byte(pemCerts))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	derPrivateKey, _ := pem.Decode([]byte(pemPrivateKey))
+	privKey, err := signedexchange.ParsePrivateKey(derPrivateKey.Bytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	certUrl, _ := url.Parse("https://example.com/cert.msg")
+	validityUrl, _ := url.Parse("https://example.com/resource.validity")
+	s := &signedexchange.Signer{
+		Date:        now,
+		Expires:     now.Add(1 * time.Hour),
+		Certs:       certs,
+		CertUrl:     certUrl,
+		ValidityUrl: validityUrl,
+		PrivKey:     privKey,
+		Rand:        zeroReader{},
+	}
+	if err := i.AddSignatureHeader(s); err != nil {
+		t.Fatal(err)
+	}
+
+	var buf bytes.Buffer
+	signedexchange.WriteExchangeFile(&buf, i)
+	got := strings.TrimSpace(base64.StdEncoding.EncodeToString(buf.Bytes()))
+	want := strings.TrimSpace(base64Result)
```

Comment:
Checking against a base64-encoded version of the CBOR data won't make it easy to diagnose test failures. You might instead use https://godoc.org/github.com/ugorji/go/codec#Decoder.Decode to decode to an interface and then `fmt` that with `%v` to get a text representation of the cbor, which will have comprehensible assertion failures.

Reply:
Done by https://github.com/WICG/webpackage/pull/119/commits/38c496415ca945dd369fcf3576f2a5416d63f9fb